src/Features/Core/Portable/AddImport/AbstractAddImportFeatureService.cs

```diff
@@ -321,40 +313,33 @@ private static bool IsHostOrRemoteWorkspace(Project project)
             return result.ToImmutableAndFree();
         }
 
-        private static async Task WaitForTasksAsync(
-            ArrayBuilder<Reference> allSymbolReferences,
+#pragma warning disable VSTHRD200 // Use "Async" suffix for async methods
+        private static async Task ProcessReferencesTask(
+            ConcurrentQueue<Reference> allSymbolReferences,
             int maxResults,
-            HashSet<Task<ImmutableArray<SymbolReference>>> findTasks,
-            CancellationTokenSource nestedTokenSource,
-            CancellationToken cancellationToken)
+            CancellationTokenSource linkedTokenSource,
+            Task<ImmutableArray<SymbolReference>> task)
         {
-            try
+            // Wait for either the task to finish, or the linked token to fire.
+            var finishedTask = await Task.WhenAny(task, Task.Delay(Timeout.Infinite, linkedTokenSource.Token)).ConfigureAwait(false);
+            if (finishedTask == task)
             {
-                while (findTasks.Count > 0)
-                {
-                    // Keep on looping through the 'find' tasks, processing each when they finish.
-                    cancellationToken.ThrowIfCancellationRequested();
-                    var doneTask = await Task.WhenAny(findTasks).ConfigureAwait(false);
-
-                    // One of the tasks finished.  Remove it from the list we're waiting on.
-                    findTasks.Remove(doneTask);
+                var result = await task.ConfigureAwait(false);
+                AddRange(allSymbolReferences, result, maxResults);
 
-                    // Add its results to the final result set we're keeping.
-                    AddRange(allSymbolReferences, await doneTask.ConfigureAwait(false), maxResults);
-
-                    // Once we get enough, just stop.
-                    if (allSymbolReferences.Count >= maxResults)
+                if (allSymbolReferences.Count >= maxResults)
+                {
+                    try
+                    {
+                        linkedTokenSource.Cancel();
+                    }
+                    catch (ObjectDisposedException)
                     {
-                        return;
                     }
                 }
             }
-            finally
-            {
-                // Cancel any nested work that's still happening.
-                nestedTokenSource.Cancel();
-            }
         }
```

Comment:
thrust of hte change.  the prior approach was n^2.  looping over a list of tasks (N), removing a single one whne finished, then running again with n-1, over and over again.  With large task lists (like when we're searching a ton of metadta dlls) this was a lot of garbage arrays for Task.WhenAny.

Now, we just let all tasks run, adding the results of the first that finish to 'allSymbolReferences' and once that reaches enough, we cancel any still in flight.